### PR TITLE
feat(core): support AbortSignal in readFile

### PR DESCRIPTION
**Pull request by benjamingr**
See discussion in https://github.com/denoland/deno/issues/10181

cc @lucacasonato

### CODE REVIEW

File: runtime/js/12_io.js
Comment (benjamingr) [reply runtime/js/12_io.js]: oh yeah

**Comment by benjamingr** [comment]
Let me know if there is anything else I should change here - if this is fine I'll go ahead and open PRs for some of the other APIs

**Review by lucacasonato — COMMENTED** [review]
Awesome! The implementation looks great. Just confirming the semantics here:

1. Initiating a readFile with an already aborted abort signal immediately (no microtask delay) rejects the returned promise.
2. Aborting a readFile that is completed already (returned promise is resolved) does nothing.
3. If the promise returned from readFile is still pending and the operation is aborted, the readFile call will always reject with an abort error, even if the read completes in the same event loop cycle.

Also we'll have to wait with landing until the 1.12 merge window opens in a couple of days. I want to get another review from @bartlomieju before merging though.

**Comment by benjamingr** [comment]
> Initiating a readFile with an already aborted abort signal immediately (no microtask delay) rejects the returned promise.

That sounds very reasonable to me and is "regular" promise semantics - the promise should be rejected _immediately_ but reacting to that rejection will always happen after a microtick. (That is, `await` and `then` are the part that wait a microtick and promise creation is typically synchronous if possible).

> Aborting a readFile that is completed already (returned promise is resolved) does nothing.

I think the behaviour should be the same as `fetch` so this is also the behaviour we've picked for Node.js - but if there are compelling alternatives I don't feel particularly strongly about it.

> If the promise returned from readFile is still pending and the operation is aborted, the readFile call will always reject with an abort error, even if the read completes in the same event loop cycle.

That's also my expectation though I'd find returning the fulfilled promise also acceptable - cancellation is "best effort" semantics. 

-----

> Also we'll have to wait with landing until the 1.12 merge window opens in a couple of days. 

That sounds fine to me - I'll wait for that and then follow up with other APIs then.

File: runtime/js/12_io.js
Comment (lucacasonato) [inline runtime/js/12_io.js:112]: Oh, didn't catch this the first time: `readAll` is a public API (a deprecated one, but still public) at `Deno.readAll`. d.ts should be changed, or the public option should not get the same semantics.

Comment (benjamingr) [reply runtime/js/12_io.js:112]: I'm fine with either - which of those would you prefer?

Comment (lucacasonato) [reply runtime/js/12_io.js:112]: Let's not expose it publically - makes it easier to change later.

Comment (benjamingr) [reply runtime/js/12_io.js:112]: Ok, does that mean just not document it (no changes) or should it be guarded (with a symbol or some-such)?

Comment (lucacasonato) [reply runtime/js/12_io.js:112]: Maybe a `readFileInner` and a `readFIle` would work?

Comment (benjamingr) [reply runtime/js/12_io.js]: Done

**Review by lucacasonato — APPROVED** [review]
LGTM. Requesting additional reviews from @ry, @piscisaureus, @kitsonk, or @bartlomieju because it touches a public API.

**Review by ry — APPROVED** [review]
LGTM - thank you @benjamingr

**Comment by benjamingr** [comment]
Thanks, I will follow up with PRs for the other APIs :)

**Pull request merged**
